src/main/java/hudson/plugins/git/extensions/impl/Credentials.java

```diff
@@ -0,0 +1,160 @@
+            if (project == null || !project.hasPermission(Item.CONFIGURE)) {
+                return FormValidation.ok();
+            }
+
+            value = Util.fixEmptyAndTrim(value);
+            if (value == null) {
+                return FormValidation.ok();
+            }
+
+            url = Util.fixEmptyAndTrim(url);
+            if (url == null)
+            // not set, can't check
+            {
+                return FormValidation.ok();
+            }
+
+            if (url.indexOf('$') >= 0)
+            // set by variable, can't check
+            {
+                return FormValidation.ok();
+            }
+
+            StandardCredentials credentials = lookupCredentials(project, value, url);
+
+            if (credentials == null) {
+                // no credentials available, can't check
+                return FormValidation.warning("Cannot find any credentials with id " + value);
+            }
+
+            // TODO check if this type of credential is acceptable to the Git client or does it merit warning the user
+
+            return FormValidation.ok();
+        }
+
+        public FormValidation doCheckUrl(@AncestorInPath Item project,
+                                         @QueryParameter String credentialsId,
+                                         @QueryParameter String value) throws IOException, InterruptedException {
+
+            if (project == null || !project.hasPermission(Item.CONFIGURE)) {
+                return FormValidation.ok();
+            }
+
+            String url = Util.fixEmptyAndTrim(value);
+            if (url == null)
+                return FormValidation.error("Please enter Git repository.");
+
+            if (url.indexOf('$') >= 0)
+                // set by variable, can't validate
+                return FormValidation.ok();
+
+            // get git executable on master
+            final EnvVars environment = new EnvVars(System.getenv()); // GitUtils.getPollEnvironment(project, null, launcher, TaskListener.NULL, false);
+
+            GitClient git = Git.with(TaskListener.NULL, environment)
+                    .using(GitTool.getDefaultInstallation().getGitExe())
+                    .getClient();
+            git.addDefaultCredentials(lookupCredentials(project, credentialsId, url));
+
+            // attempt to connect the provided URL
+            try {
+                git.getHeadRev(url, "HEAD");
+            } catch (GitException e) {
+                return FormValidation.error(Messages.UserRemoteConfig_FailedToConnect(e.getMessage()));
+            }
+
+            return FormValidation.ok();
+        }
+
+        private static StandardCredentials lookupCredentials(Item project, String credentialId, String uri) {
+            return (credentialId == null) ? null : CredentialsMatchers.firstOrNull(
+                        CredentialsProvider.lookupCredentials(StandardCredentials.class, project, ACL.SYSTEM,
```

Comment:
Abusing the user to look up credentials as... if the project is being run as a user other than ACL.SYSTEM then using this user to look up will cause issues